### PR TITLE
Allow plain functions to be used as components

### PR DESCRIPTION
**Pull request by BinaryMuse**
Often, a component only needs to take in props/children and render something, and does not maintain any internal state. In those instances, it can be useful to define the component as a function of props and
children.

This adds `etch.stateless(renderFn)` that implements a component that renders `renderFn` with props and children during `initialize` and `update`.

@kuychaco has brought up the excellent point that `etch.stateless` isn't the best name for this; I'm totally open to suggestions

Signed-off-by: Katrina Uychaco <kuychaco@github.com>

### CODE REVIEW

**Comment by BinaryMuse** [comment]
I've been looking at this a bit more and I think it can be improved. I would like to support the following use cases, in order of importance:

1. Use plain functions as components.

  In this scenario, a function that takes props and children and returns a `VNode` could be a valid component type; for instance:

  ```javascript
  function PlainFunctionComponent({thing}, children) {
    return <div>{thing} - {children}</div>
  }

  // ...

  render() {
    return (
      <div>
        <PlainFunctionComponent thing="yep">
          <span>children!</span>
        </PlainFunctionComponent>
      </div>
    )
  }
  ```

  This simplifies many cases where you just want to render something static based on props and need no internal state tracking in the component.

2. Support plain function components as top-level components.

  This would allow users to use such functions as entry-points into Etch-based views. It's less clear to me how something like this would look/work (or if it's really that necessary at all). I could see both the following cases being useful:

  1. Just rendering some static Etch view directly from a `VNode`:

    ```javascript
    let element = etch.render(functionReturningVNode(props))
    document.body.appendChild(element)
    ```

  2. Using a plain function to render a simple view that contains stateful components:

    ```javascript
    function MyComponent({name}) {
      return <StatefulComponent name={name} />
    }

    let component = etch.render(MyComponent, {name: 'Alice'})
    document.body.appendChild(component.element)
    // later
    component.update({name: 'Bob'})
    ```

File: src/dom.js [hunk 1]
```diff
@@ -16,6 +17,9 @@ import ComponentWidget from './component-widget'
 // particular library. For more information, see `./component-widget.js`.
 export default function dom (tag, properties, ...children) {
   if (typeof tag === 'function') {
+    if (!tag.prototype.render) {
+      tag = fromFunction(tag)
+    }
```
Comment (BinaryMuse) [inline src/dom.js]: Not 100% sure if this is the best method to detect if we're using a plain function instead of a component constructor

**Comment by BinaryMuse** [comment]
I decided to drop case 2 detailed above. Since Etch components are objects with a consistent interface, it didn't seem to make sense to force plain functions in at the top-level. I did export a new `etch.render` that can be used to render simple VNodes directly to elements.

**Comment by BinaryMuse** [comment]
Okay, I think this is ready for some 👀. /cc @nathansobo

**Comment by joshaber** [comment]
😍

**Comment by nathansobo** [comment]
It seems like there are a few different issues in play here.

1) **Using functions as child components.** Instead of passing a constructor as a "tag name", you want to pass a function returning JSX. This may be a matter of detecting whether the return value is a JSX expression or a component object and doing the right thing. Unfortunately, to conform to people's expectations about how constructors work, we always call child component functions with `new` assuming they're constructors. Is there any way to definitively tell whether something is actually a constructor, perhaps by checking prototypes or something? That would help the efficiency of such a polymorphic move.

2) **Easily defining component objects via functions**. This seems like a library or helper function to me that just takes the function and does the work of setting up a simple component that calls `etch.initialize` and `etch.update` at the right times, something akin to your `etch.stateless` idea. I'm not fully convinced this is something we should include in etch core at this time, but I'm not 100% opposed and could be convinced.

3) More amorphous concerns about working with JSX and functions that return it. Given a JSX expression, how do I render it to DOM nodes? Given a function returning JSX, what else useful can I do with it outside the framework of the Etch component lifecycle methods. For this, I originally exported some of the primitives we use internally such as `diff`, `patch`, `render` etc. This may be something we could explore again, but I want to see use cases to inform the decision.

**Comment by BinaryMuse** [comment]
Item 2 isn't very important to me; originally, `etch.stateless` was simply a way to support item 1. Now it's used to get around the fact that I'm not sure what the best way is to determine if a function is a constructor or not; thus I stopped exporting it from Etch (it's only used internally). If we can detect if a function isn't a constructor (and just returns JSX), then as per your first point, we could probably just call the function and do away with `etch.stateless` completely.

**Comment by jeancroy** [comment]
Would it make any sens to simply have a `etch.fromRenderFunction(fn)` helper that scaffold a basic life cycle component and use the function to render it?

I understand there's design intent to hide etch as a implementation detail, but it seems the amount of boilerplate copy paste may be inconvenient in some case. Also it seems that because of virtual dom there's always some amount of state to maintain, live object VS jsx string.

**Comment by BinaryMuse** [comment]
@nathansobo 

> Is there any way to definitively tell whether something is actually a constructor, perhaps by checking prototypes or something? That would help the efficiency of such a polymorphic move.

BTW this is exactly what this PR is doing now — it creates a Widget from a plain function if that function doesn't have a `render` method on its prototype.

File: src/component-helpers.js [hunk 2]
```diff
@@ -103,9 +103,7 @@ export function updateSync (component) {
   let newDomNode = patch(component.element, diff(oldVirtualElement, newVirtualElement))
   refsStack.pop()
   component.virtualElement = newVirtualElement
-  if (newDomNode !== oldDomNode) {
-    throw new Error("etch does not support changing the root DOM node type of a component")
-  }
+  component.element = newDomNode
```
Comment (BinaryMuse) [inline src/component-helpers.js]: Superseded by #20

**Comment by BinaryMuse** [comment]
Chatted with @nathansobo a bunch today and we ultimately decided to back out of this, in favor of a solution like the one implemented in 88e756feff3c33769dc47fc53e193b49d8e53260 due to complexities around ensuring that the right version of etch is used to update functional components. This solution will live in userland, at least for now.

**Comment by BinaryMuse** [comment]
@nathansobo we talked about moving this to userland, but doesn't that result in the userland module that provides the `stateless` function using `initialize` and `update` from its version of etch, and the user's module using the `etch.dom` constructors from its version of etch?

**Pull request closed without merging**